compiler/rustc_ty/src/needs_drop.rs

```diff
@@ -94,15 +94,21 @@ where
                     _ if component.is_copy_modulo_regions(tcx.at(DUMMY_SP), self.param_env) => (),
 
                     ty::Closure(_, substs) => {
-                        for upvar_ty in substs.as_closure().upvar_tys() {
-                            queue_type(self, upvar_ty);
+                        // Ensure that the upvar tys has been inferred
+                        if substs.as_closure().is_valid() {
```

Comment:
I think this could be `queue_type(substs.as_closure().tupled_upvars_ty())`, no?

Reply:
(same with generators)

Reply:
done